File: ios/SpentyAI/SpentyAI/Features/Reports/ReportsRepository.swift
import Foundation

struct PeriodsResponse: Codable {
    let periods: [ReportPeriod]
}

struct CategoriesResponse: Codable {
    let categories: [ReportCategory]
}

final class ReportsRepository {

    static let shared = ReportsRepository()
    private init() {}

    // MARK: - Summary

    func getSummary(from: Date, to: Date) async throws -> ReportSummary {
        let query = dateQuery(from: from, to: to)
        return try await APIClient.shared.get(APIEndpoints.reportsSummary + query)
    }

    // MARK: - Period Breakdown

    func getPeriods(from: Date, to: Date) async throws -> [ReportPeriod] {
        let query = dateQuery(from: from, to: to)
        let response: PeriodsResponse = try await APIClient.shared.get(APIEndpoints.reportsByPeriod + query)
        return response.periods
    }

    // MARK: - Category Breakdown

    func getCategories(from: Date, to: Date, type: String = "expense") async throws -> [ReportCategory] {
        let query = dateQuery(from: from, to: to) + "&transaction_type=\(type)"
        let response: CategoriesResponse = try await APIClient.shared.get(APIEndpoints.reportsByCategory + query)
        return response.categories
    }

    // MARK: - Income vs Expense

    func getIncomeExpense(from: Date, to: Date) async throws -> IncomeExpenseResponse {
        let query = dateQuery(from: from, to: to)
        return try await APIClient.shared.get(APIEndpoints.reportsIncomeExpense + query)
    }

    // MARK: - Export

    func exportCSV(from: Date, to: Date) async throws -> Data {
        let query = dateQuery(from: from, to: to)
        return try await APIClient.shared.getRaw(APIEndpoints.reportsExportCSV + query)
    }

    func exportPDF(from: Date, to: Date) async throws -> Data {
        let query = dateQuery(from: from, to: to)
        return try await APIClient.shared.getRaw(APIEndpoints.reportsExportPDF + query)
    }

    // MARK: - Helpers

    private func dateQuery(from: Date, to: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withFullDate]
        let fromStr = formatter.string(from: from)
        let toStr = formatter.string(from: to)
        return "?start_date=\(fromStr)&end_date=\(toStr)"
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/reports/ReportsRepository.kt
package com.spentyai.app.features.reports

import com.spentyai.app.core.models.*
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ReportsRepository(private val apiClient: ApiClient) {

    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    suspend fun getSummary(from: Date, to: Date): ApiResult<ReportSummary> {
        val fromStr = dateFormatter.format(from)
        val toStr = dateFormatter.format(to)
        return apiClient.safeApiCall {
            apiClient.endpoints.getReportsSummary(fromStr, toStr)
        }
    }

    suspend fun getPeriods(from: Date, to: Date): ApiResult<List<ReportPeriod>> {
        val fromStr = dateFormatter.format(from)
        val toStr = dateFormatter.format(to)
        return apiClient.safeApiCall {
            apiClient.endpoints.getReportsByPeriod(fromStr, toStr)
        }.map { it.periods }
    }

    suspend fun getCategories(from: Date, to: Date, type: String = "expense"): ApiResult<List<ReportCategory>> {
        val fromStr = dateFormatter.format(from)
        val toStr = dateFormatter.format(to)
        return apiClient.safeApiCall {
            apiClient.endpoints.getReportsByCategory(fromStr, toStr, type)
        }.map { it.categories }
    }

    suspend fun getIncomeExpense(from: Date, to: Date): ApiResult<IncomeExpenseResponse> {
        val fromStr = dateFormatter.format(from)
        val toStr = dateFormatter.format(to)
        return apiClient.safeApiCall {
            apiClient.endpoints.getReportsIncomeExpense(fromStr, toStr)
        }
    }

    suspend fun exportCSV(from: Date, to: Date): ApiResult<ByteArray> {
        val fromStr = dateFormatter.format(from)
        val toStr = dateFormatter.format(to)
        return apiClient.safeApiCall {
            apiClient.endpoints.exportReportCSV(fromStr, toStr)
        }.map { it.bytes() }
    }

    suspend fun exportPDF(from: Date, to: Date): ApiResult<ByteArray> {
        val fromStr = dateFormatter.format(from)
        val toStr = dateFormatter.format(to)
        return apiClient.safeApiCall {
            apiClient.endpoints.exportReportPDF(fromStr, toStr)
        }.map { it.bytes() }
    }
}
